src/modules/sdbus/progress.c

```diff
@@ -0,0 +1,171 @@
+/************************************************************\
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* progress.c - integrate sd_bus with the flux reactor
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <poll.h>
+#include <errno.h>
+#include <time.h>
+#include <flux/core.h>
+#include <math.h>
+#include <systemd/sd-bus.h>
+
+#include "progress.h"
+
+struct progress {
+    flux_watcher_t *in;
+    flux_watcher_t *out;
+    flux_watcher_t *tmout;
+    sd_bus *bus;
+    flux_t *h;
+    progress_recv_f recv_cb;
+    void *recv_arg;
+};
+
+/* sd_bus_get_timetout () sets usec to CLOCK_MONOTONIC absolute time,
+ * with UINT64_MAX indicating "no timeout".  See notes in sd_bus_get_fd(3).
+ */
+static int get_relative_timeout (sd_bus *bus, double *timeout)
+{
+    uint64_t usec;
+    int e;
+
+    if ((e = sd_bus_get_timeout (bus, &usec)) < 0) {
+        errno = -e;
+        return -1;
+    }
+    if (usec == UINT64_MAX)
+        *timeout = -1;  // no timeout
+    else if (usec == 0)
+        *timeout = 0;   // immediate timeout
+    else {
+        struct timespec ts;
+        double now;
+        if (clock_gettime (CLOCK_MONOTONIC, &ts) < 0)
+            return -1;
+        now = floor (1E6*ts.tv_sec + 1E-3*ts.tv_nsec);  // to usec (round down)
+        *timeout = 1E-6*(usec - now);
+    }
+    return 0;
+}
+
+void progress_stop (struct progress *prog)
+{
+    flux_watcher_stop (prog->in);
+    flux_watcher_stop (prog->out);
+    flux_watcher_stop (prog->tmout);
+}
+
+void progress_start (struct progress *prog)
+{
+    int events;
+    double timeout;
+
+    if ((events = sd_bus_get_events (prog->bus)) < 0) {
+        errno = -events;
+        flux_log_error (prog->h, "failed to get sd_bus progress events");
+        goto fatal;
+    }
+    if (get_relative_timeout (prog->bus, &timeout) < 0) {
+        flux_log_error (prog->h, "failed to get sd_bus progress timeout");
+        goto fatal;
+    }
+
+    //flux_log (prog->h, LOG_DEBUG, "bus timeout=%.3fs events=0x%x", timeout, events);
+
+    if ((events & POLLIN))
+        flux_watcher_start (prog->in);
+    if ((events & POLLOUT))
+        flux_watcher_start (prog->out);
+    flux_timer_watcher_reset (prog->tmout, timeout, 0.);
+    if (timeout >= 0.)
+        flux_watcher_start (prog->tmout);
+    return;
+fatal:
+    flux_reactor_stop_error (flux_get_reactor (prog->h));
+}
```

Comment:
I should also note, I don't think events == 0 and timeout != 0 is possible, but worthwhile to add a check and log if it happens?

Reply:
[sd_bus_get_fd(3)](https://manpages.ubuntu.com/manpages/focal/man3/sd_bus_get_fd.3.html) describes what I was trying to do here:
>  These three functions are useful to hook up a bus connection
       object with an external or manual event loop involving poll() or
       a similar I/O polling call. Before each invocation of the I/O
       polling call, all three functions should be invoked: the file
       descriptor returned by sd_bus_get_fd() should be polled for the
       events indicated by sd_bus_get_events(), and the I/O call should
       block for that up to the timeout returned by
       sd_bus_get_timeout(). After each I/O polling call the bus
       connection needs to process incoming or outgoing data, by
       invoking sd_bus_process(3).

This is accomplished by setting up flux POLLIN and POLLOUT watchers on the file descriptor, and a flux timer watcher for the timeout, and then whenever any of those is ready, calling `sd_bus_process()` to let the sdbus API event loop make progress and possibly get a received message for flux to process.

I don't think the order of those calls is significant.  We have to asynchronously block waiting for any of POLLIN, POLLOUT, or the timeout to be ready, and that blocking doesn't occur until the reactor callback exits.

Reply:
I'll have to play with this a bit later, but my recollection is this bit me a few times.  As I look through `libsdprocess` some more, I do not setup a timer watcher, only a fd watcher, so checking for timeout == 0 matters a lot more for that code.

Reply:
Yes the main thing is that the flux reactor cannot be allowed to block if the sdbus reactor has a pending timeout.  If the timeout is 0, that is treated like any other timeout - the timer watcher becomes ready when the timeout expires, immediately in this case.

Reply:
In practice, I don't think I've ever seen events set to anything but 1 (POLLIN).

However I don't think there's any requirement that timeout must be 0 if events is 0.  I think your concern must be about events = 0 (no events) and timeout = -1 (no timeout)?  I assumed sdbus wouldn't return that combination but if it did I don't think we would ever call the progress functions again.  Hmm, at minimum we should treat that as a fatal error if it occurs.  And if that ever happens, we might want to convert this code to use prep/check/idle watchers.  I'll have to think about that one - it might be better to use prep/check/idle anyway.

Reply:
clarification: by timeout == -1 I mean the post-processed timeout in progress.c.  `sd_bus_get_timeout()` returns UINT64_MAX for "no timeout" and that is turned into -1 (compatible with the flux timer watcher) in `get_relative_timeout()`.

Reply:
> Hmm, at minimum we should treat that as a fatal error if it occurs. 

This is where there are subtleties in `systemd` that I think neither of us truly know :P  I'm wondering if in `libsdprocess` we see events == 0 b/c of my unique setup and it simply isn't possible under yours.

Perhaps a couple of "we assume this is impossible" fatal error checks would be worthwhile and if it ends up biting us later, then code can be re-worked?

Reply:
Oh maybe events is always 1 because I always have a subscription going.

> Perhaps a couple of "we assume this is impossible" fatal error checks would be worthwhile and if it ends up biting us later, then code can be re-worked?

Wholeheartedly agree!